juju/juju
# Handle AWS bindings placements

## Description of change

The following changes attempt to handle both space constraints and
binding placements when looking for subnets in availability zones.

Previously the code looked only for spaces, but binding placements
although similar aren't spaces, so the code path required to do the
a lookup was failing for this.

In doing the changes, I also refactored the code to highlight the
similarities (random selection and filtering) and the differences, the
VPC ID and the lack of VPC ID selection.

Additionally, I've added some unit tests to ensure that we correctly test
at least some of the environ code, as the time of writing we have very
little to no real unit tests for individual functionality for pure unit
level testing.

## QA steps

Assuming you're in `eu-west-1a`, bootstrap a controller and add a space.

```sh
juju bootstrap aws aws-test
juju add-space space-1 172.31.32.0/20
juju deploy cs:percona-cluster-290 --bind "space-1"
juju show-machine 0
```

The following should now be in `eu-west-1b`

```yaml
model: default
machines:
  "0":
    juju-status:
      current: pending
      since: 09 Jun 2020 11:14:02+01:00
    dns-name: 34.250.73.229
    ip-addresses:
    - 34.250.73.229
    - 172.31.42.149
    instance-id: i-04cd9f5780d8f72c6
    machine-status:
      current: running
      message: running
      since: 09 Jun 2020 11:14:25+01:00
    modification-status:
      current: idle
      since: 09 Jun 2020 11:14:02+01:00
    series: bionic
    hardware: arch=amd64 cores=2 cpu-power=700 mem=1024M root-disk=8192M availability-zone=eu-west-1b
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1880714
https://bugs.launchpad.net/juju/+bug/1881431
